### PR TITLE
Support new Self type as an annotation.

### PR DESCRIPTION
**Pull request by coady**
## Description
`typing.Self` is a supported annotation, for types and interfaces.

```python
from strawberry.types import Self

@strawberry.type
class Node:
    @strawberry.field
    def field(self) -> Self:
        return self
```

## Types of Changes
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #2292

## Checklist
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the release file! Unfortunately it does seem to be invalid. Make sure it looks like the following example:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to A. Coady for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/types/types.py [hunk 1]
```diff
@@ -23,6 +23,11 @@
 
     from strawberry.field import StrawberryField
 
+try:
+    from typing import Self  # type: ignore
+except ImportError:
+    Self = TypeVar("Self")
```
Comment (patrick91) [inline strawberry/types/types.py:29]: what about importing it from typing extensions?

File: strawberry/types/types.py [hunk 2]
```diff
@@ -44,6 +46,11 @@ class TypeDefinition(StrawberryType):
         default_factory=dict
     )
 
+    def __post_init__(self):
+        for index, field in enumerate(self.fields):
+            if isinstance(field.type, StrawberryType) and field.type.has_generic(Self):
+                self.fields[index] = field.copy_with({Self: self.origin})  # type: ignore
```
Comment (patrick91) [inline strawberry/types/types.py:52]: can you add a comment explaining why this is needed? 😊

File: tests/schema/test_generics.py [hunk 3]
```diff
@@ -815,6 +817,37 @@ def books(self) -> List[Book]:
...
+      fields: [INode!]!
+    }
+
+    type Node implements INode {
+      field: Node!
+      fields: [Node!]!
+    }
+    """
+
+    assert str(schema) == textwrap.dedent(expected_schema).strip()
```
Comment (patrick91) [inline tests/schema/test_generics.py:848]: can we also add a test making sure we can fetch fields? :)

**Review by patrick91 — COMMENTED** [review]
This looks neat! 

What are your use cases for using Self?

**Comment by coady** [comment]
> What are your use cases for using Self?

I think the general use case is that using `Self` will become standard in annotated Python going forward, so the expectation that it works with Strawberry will increase.

But my use case is pretty much the test example:
* an interface `Foo`, with field(s) `bar` which returns `Foo`
* implementations `*Foo` override `bar` to return their compatible `*Foo`
* the `Foo` resolver uses `type(self)`, so the implementations could just inherit and reuse its resolver
* except... the resolvers needs to wrapped or duplicated anyway because they have to register `*Foo` somehow

**Review by patrick91 — APPROVED** [review]
Nice work!

**Pull request merged**